### ISSUE TITLE
x-ent softmax with bias: dimension mismatch in x,y_idx arguments

### ISSUE THREAD

**Issue by lucasb-eyer** [Theano/Theano#2464]
This is the tutorial's code for x-entropy:

```
p_y_given_x = T.nnet.softmax(T.dot(X, W) + b)
loglik = T.log(p_y_given_x)[T.arange(y.shape[0]), y]
```

Reading the above code, and `p_y_given_x` of shape `(n_batch, n_class)`, I expect `loglik` to be of shape `(y.shape[0], 1)`, (which it is in pure numpy code) but instead I get an error because the code got replaced by `GpuCrossentropySoftmaxArgmax1HotWithBias` which wants to have `y.shape[0] == p_y_given_x.shape[0]`.

I think [that test](https://github.com/Theano/Theano/blob/master/theano/sandbox/cuda/nnet.py#L117) should be relaxed to only verify `<=` and not `==` , and then pass dim-0 of `y_idx` instead of `x` [here](https://github.com/Theano/Theano/blob/master/theano/sandbox/cuda/nnet.py#L183). Unfortunately, I know too little theano internals to try this myself.

The use-case is to ignore the padded entries of the last minibatch when batchsize doesn't fit data-size nicely.

**Comment by lamblin** [comment]
Do you mean that, in your case, `y.shape[0]` is different from `n_batch`?
If that is the case, then indeed inserting the `CrossentropySoftmax...` operation is wrong, however it is hard to detect correctly at compile-time, and most of the time it is still the right thing to do.
Relaxing the test would not really be enough, we would also have to change the size of the output, and so on. It can be done (in the CPU code as well), but I'm not sure when someone will have time for it.

Meanwhile, you can work around this problem in a couple of ways:
- only compute `p_y_given_x` on valid entries: `p_y_given_x = T.nnet.softmax(T.dot(X[:y.shape[0]], W) + b)`
- pad `y` as well as `X`.

**Comment by lucasb-eyer** [comment]
Yes, that's exactly what I mean. I might try to give it a crack myself when I get time, if I ever will...

Padding `y` is what I've been doing so far, but unless I'm missing something obvious, it's quite hacky to ignore the padded entries in the computation of the minibatch's mean nll.

**Comment by lamblin** [comment]
FWIW, the issue is still there. I could reproduce it with:
```python
import numpy as np
import theano
from theano import tensor


X = tensor.matrix('X')
y = tensor.lvector('y')
W = tensor.matrix('W')
b = tensor.vector('b')

p_y_given_x = tensor.nnet.softmax(tensor.dot(X, W) + b)
loglik = tensor.log(p_y_given_x)[tensor.arange(y.shape[0]), y]

X_val = np.ones((3, 4), dtype='f')
y_val = np.array([0, 1], dtype='l')  # Note that y.shape[0] < X.shape[1]
W_val = np.ones((4, 2), dtype='f')
b_val = np.zeros((2,), dtype='f')

f = theano.function([X, y, W, b], loglik)
theano.printing.debugprint(f, print_type=1)

# Workaround
X_ = X[:y.shape[0]]
p_y_given_x_ = tensor.nnet.softmax(tensor.dot(X_, W) + b)
loglik_ = tensor.log(p_y_given_x_)[tensor.arange(y.shape[0]), y]
g = theano.function([X, y, W, b], loglik_)
print(g(X_val, y_val, W_val, b_val))
print(f(X_val, y_val, W_val, b_val))
```
A solution would be to update the optimization to use `X[:y.shape[0]]` in the first place, which should not be hard.
